Signal/src/views/OWSContactOffersCell.m

```diff
@@ -0,0 +1,236 @@
+    self.titleLabel.lineBreakMode = NSLineBreakByTruncatingTail;
+    self.titleLabel.textAlignment = NSTextAlignmentCenter;
+    [self.contentView addSubview:self.titleLabel];
+
+    self.addToContactsButton = [self
+        createButtonWithTitle:
+            NSLocalizedString(@"CONVERSATION_VIEW_ADD_TO_CONTACTS_OFFER",
+                @"Message shown in conversation view that offers to add an unknown user to your phone's contacts.")
+                     selector:@selector(addToContacts)];
+    self.addToProfileWhitelistButton = [self
+        createButtonWithTitle:NSLocalizedString(@"CONVERSATION_VIEW_ADD_USER_TO_PROFILE_WHITELIST_OFFER",
+                                  @"Message shown in conversation view that offers to share your profile with a user.")
+                     selector:@selector(addToProfileWhitelist)];
+    self.blockButton =
+        [self createButtonWithTitle:NSLocalizedString(@"CONVERSATION_VIEW_UNKNOWN_CONTACT_BLOCK_OFFER",
+                                        @"Message shown in conversation view that offers to block an unknown user.")
+                           selector:@selector(block)];
+}
+
+- (UIButton *)createButtonWithTitle:(NSString *)title selector:(SEL)selector
+{
+    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
+    [button setTitle:title forState:UIControlStateNormal];
+    [button setTitleColor:[UIColor ows_materialBlueColor] forState:UIControlStateNormal];
+    button.titleLabel.font = self.buttonFont;
+    button.titleLabel.textAlignment = NSTextAlignmentCenter;
+    [button setBackgroundColor:[UIColor colorWithRGBHex:0xf5f5f5]];
+    button.layer.cornerRadius = 5.f;
+    [button addTarget:self action:selector forControlEvents:UIControlEventTouchUpInside];
+    [self.contentView addSubview:button];
+    return button;
+}
+
++ (NSString *)cellReuseIdentifier
+{
+    return NSStringFromClass([self class]);
+}
+
+- (void)configureWithInteraction:(OWSContactOffersInteraction *)interaction;
+{
+    OWSAssert(interaction);
+
+    _interaction = interaction;
+
+    OWSAssert(
+        interaction.hasBlockOffer || interaction.hasAddToContactsOffer || interaction.hasAddToProfileWhitelistOffer);
+
+    [self setNeedsLayout];
+}
+
+- (UIFont *)titleFont
+{
+    return [UIFont ows_mediumFontWithSize:16.f];
+}
+
+- (UIFont *)buttonFont
+{
+    return [UIFont ows_regularFontWithSize:14.f];
+}
+
+- (CGFloat)hMargin
+{
+    return 10.f;
+}
+
+- (CGFloat)topVMargin
+{
+    return 5.f;
+}
+
+- (CGFloat)bottomVMargin
+{
+    return 5.f;
+}
+
+- (CGFloat)buttonVPadding
+{
+    return 5.f;
+}
+
+- (CGFloat)buttonVSpacing
+{
+    return 5.f;
+}
+
+- (void)layoutSubviews
+{
+    [super layoutSubviews];
+
+    // We're using a bit of a hack to get JSQ to layout this and the unread indicator as
+    // "full width" cells.  These cells will end up with an erroneous left margin that we
+    // want to reverse.
+    CGFloat contentWidth = self.width;
+    CGFloat left = -self.left;
```

Comment:
Slightly different version of the same hack that we're using in the unread indicator.